mesonbuild/coredata.py

```diff
@@ -876,18 +877,20 @@ def __init__(self, filenames: T.List[str]) -> None:
         self.constants = {'True': True, 'False': False}
         self.sections = {}
 
-        self.parser.read(filenames)
+        for fname in filenames:
+            self.parser.read(fname)
```

Comment:
I don't think you should change this, configparser has special treatment when reading multiple files.

Reply:
Ah, I understand now why you did that, you want paths to be relative to the machine file, and not relative to the source dir... It actually make sense... But this changes slightly the behaviour of the parser I think, because now you're evaluating constants for each file instead of after merging all files together. It is possible for example to have one machine file with only `[constants]` and another file that uses those constants, the idea being you can have common parts in one file and only switch the constants file for each arch.

I'm not sure how to make this work though....

Reply:
First I agreed with you, that the handling of constants has changed. So I added a small unittest containing three machinfiles where the constant `arch` is overwritten in the last crossfile.
```python
def test_cross_file_constants_merge(self):
        with temp_filename() as crossfile1, temp_filename() as crossfile2, temp_filename() as crossfile3:
            with open(crossfile1, 'w', encoding='utf-8') as f:
                f.write(textwrap.dedent(
                    '''
                    [constants]
                    arch = 'win64'
                    '''))
            with open(crossfile2, 'w', encoding='utf-8') as f:
                f.write(textwrap.dedent(
                    '''
                    [binaries]
                    c = arch + '-gcc'
                    '''))
            with open(crossfile3, 'w', encoding='utf-8') as f:
                f.write(textwrap.dedent(
                    '''
                    [constants]
                    arch = 'arm64'

                    [binaries]
                    g++ = arch + '-g++'
                    '''))
            config = mesonbuild.coredata.parse_machine_files([crossfile1, crossfile2, crossfile3])
            bt = mesonbuild.envconfig.BinaryTable(config.get('binaries', {}))
            self.assertEqual(bt.binaries['c'], ['arm64-gcc'])
            self.assertEqual(bt.binaries['g++'], ['arm64-g++'])
```
My expectation was that with my change the binary `c` equals to `win64-gcc` instead of `arm64-gcc`, unfortunately the unittest passed... 

So after digging into the code it was clear: Every time the configparser is called by `self.parser.read(fname)` the new machinefile is merged into the already parsed data, so at the end the behaviour is identical.

BUT this also means that if there are e.g. 2 machinefiles with binary entries with a relative path, the evaluation of the last machinefile overwrites the path from the first one.

=> still needs some rework